src/sentry/sentry_metrics/querying/units.py

```diff
@@ -85,6 +90,36 @@ class UnitsSpec:
     units: Sequence[Unit]
 
 
+@dataclass(frozen=True)
+class UnitMetadata:
+    pass
+
+
+@dataclass(frozen=True)
+class WithNoUnit(UnitMetadata):
+    pass
+
+
+@dataclass(frozen=True)
+class WithFutureUnit(UnitMetadata):
+    pass
+
+
+@dataclass(frozen=True)
+class WithUnit(UnitMetadata):
```

Comment:
Used as a way to type return values similarly to how tagged unions would work. Not really the same but the developer experience aims at being similar.